docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,2165 @@
+        ├── [Module]Config.php
+        └── [Module]DependencyProvider.php
+```
+
+**Description**
+
+Injects required dependencies to a module application layer. Typically, dependencies are [facades](#facade-design-pattern) or [plugins](#plugin). 
+
+Dependency injection is orchestrated through a `provide-add-get` structure (see example below).
+- The `provide` method is the highest level that holds only `add` calls without any additional logic.
+- The `add` method is the middle level that injects the dependencies into the dependency container using a class constant and a late-binding instantiating closure.
+- The `get` method is the lowest level that sources the dependency.
+
+**Conventions**
+- All class constants MUST be `public` (to decrease conflicts in definition for being a public API class).
+- Each dependency MUST be defined using a class constant.
+- Setting dependency using the `container::set()` MUST be paired with late-binding closure definition to decouple instantiation.
+- Dependencies MUST be wired through the target layer corresponding inherited method:
+```php
+public function provideCommunicationLayerDependencies(Container $container)
+public function provideBusinessLayerDependencies(Container $container)
+public function providePersistenceLayerDependencies(Container $container)
+public function provideDependencies(Container $container)
+public function provideBackendDependencies(Container $container)
+public function provideServiceLayerDependencies(Container $container)
+```
+- Only three type of methods CAN be defined, either `provide`, `get`, or `add`.
+```php
+public function provide*Dependencies(Container $container)
+```
+- `Provide` methods MUST call their parent `provide` method.
+- `Provide` methods MUST only call `add` methods.
+
+```php
+protected function add*(Container $container)
+```
+- `Add` methods MUST only introduce one dependency to the `Container` (a plugin-stack is considered one dependency in this respect).
+
+```php
+protected function get[Dependency](Container $container)
+protected function get[PluginInterfaceName]Plugins(Container $container)
+```
+
+<details><summary markdown='span'>Additional Conventions for Project Development</summary>
+- Constants can have any access modifier, they are not limited to `public`. <br/>
+- Methods can have any access modifier. <br/>
+- `Provide` methods can call other methods than `add`. <br/>
+- `Add` methods can introduce more dependency to the `Container`.<br/>
+- The `Container $container` argument can be introduced on need-to-have basis instead of being always mandatory.
+</details>
+
+<details><summary markdown='span'>Additional Conventions for Module Development</summary>
+- [Facades](#facade-design-pattern) MUST be wrapped into a [Bridge](#bridge) to avoid coupling another module.<br/>
+- [Plugins](#plugin) MUST NOT be wired on module level.<br/>
+- [Plugin](#plugin) defining methods MUST be tagged with `@api` tag as they are considered module API. <br/>
```

Comment:
Dependency constant value must be the same as constant name

Reply:
constants must have a prefix that describe its purpose. E.g. **FACADE**_STORE, **PLUGINS**_EXAMPLE, etc.